addons/l10n_vn_edi_viettel/models/account_move.py

```diff
@@ -0,0 +1,766 @@
+            'l10n_vn_edi_invoice_state': False,
+        })
+        # Cleanup the files as well. They will still be available in the chatter.
+        cancelled_sinvoices.l10n_vn_edi_sinvoice_xml_file_id.unlink()
+        cancelled_sinvoices.l10n_vn_edi_sinvoice_pdf_file_id.unlink()
+        cancelled_sinvoices.l10n_vn_edi_sinvoice_file_id.unlink()
+        return res
+
+    def _l10n_vn_edi_generate_invoice_json(self):
+        """ Return the dict of data that will be sent to the api in order to create the invoice. """
+        # We leave the summarized information computation to SInvoice.
+        self.ensure_one()
+        # This MUST match chronologically with the sequence they generate on their system, which is why it is set to now.
+        self.l10n_vn_edi_issue_date = fields.Datetime.now()
+        json_values = {}
+        self._l10n_vn_edi_add_general_invoice_information(json_values)
+        self._l10n_vn_edi_add_buyer_information(json_values)
+        self._l10n_vn_edi_add_seller_information(json_values)
+        self._l10n_vn_edi_add_payment_information(json_values)
+        self._l10n_vn_edi_add_item_information(json_values)
+        self._l10n_vn_edi_add_tax_breakdowns(json_values)
+        return json_values
+
+    def _l10n_vn_edi_add_general_invoice_information(self, json_values):
+        """ General invoice information, such as the model number, invoice symbol, type, date of issues, ... """
+        self.ensure_one()
+        invoice_data = {
+            'transactionUuid': str(uuid.uuid4()),
+            'invoiceType': self.l10n_vn_edi_invoice_symbol.invoice_template_id.template_invoice_type,
+            'templateCode': self.l10n_vn_edi_invoice_symbol.invoice_template_id.name,
+            'invoiceSeries': self.l10n_vn_edi_invoice_symbol.name,
+            # This timestamp is important as it is used to check the chronological order of Invoice Numbers.
+            # Since this xml is generated upon posting, just like the invoice number, using now() should keep that order
+            # correct in most case.
+            'invoiceIssuedDate': self._l10n_vn_edi_format_date(self.l10n_vn_edi_issue_date),
+            'currencyCode': self.currency_id.name,
+            'adjustmentType': '1',  # 1 for original invoice, which is the case during first issuance.
+            'paymentStatus': self.payment_state in {'in_payment', 'paid'},
+            'cusGetInvoiceRight': True,  # Set to true, allowing the customer to see the invoice.
+            'validation': 1,  # Set to 1, SInvoice will validate tax information while processing the invoice.
+        }
+
+        # When invoicing in a foreign currency, we need to provide the rate, or it will default to 1.
+        if self.currency_id.name != 'VND':
+            invoice_data['exchangeRate'] = self.env['res.currency']._get_conversion_rate(
+                from_currency=self.company_currency_id,
+                to_currency=self.currency_id,
```

Comment:
You never check the currency of the company when allowing or not the user to send to SInvoice. What if the company's currency is USD for example?

Reply:
You are right. They always expect conversion from invoice currency to VND, this is an oversight.
Even if the company currency is VND, 

I don't think we need to block companies using USD to issue invoices, but we should ensure that VND is at least enabled and has a conversion rate != 1 if it's not the company currency I guess.